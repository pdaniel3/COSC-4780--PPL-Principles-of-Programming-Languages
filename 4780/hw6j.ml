(* Phil Daniels Homework 6 *)

(* PROOFS -------------------------------------------------------
PROOFS
6.2a First we know that both sides of the equality have the 
type of command, so we can apply extensionality. Take some
arbitrary store s

[[C1;(C2;C3):comm]](s) = [[(C1;C2);C3:comm]](s)

Starting from the left...

[[C1;(C2;C3):comm]](s) = [[C2;C3:comm]]([[C1:comm]](s)) 
                       = [[C3:comm]]([[C2:comm]]([[C1:comm]](s)))
                       = [[C3:comm]]([[C1;C2:comm]](s)) 
                       = [[(C1;C2);C3:comm]](s)

6.2c. So I ran out of time, so this is an informal proof...
Because both left and right have the same type (comm), we 
can apply extensionality. These two statements being equal 
makes sense intuitively. E : boolExp so it can be either 
true or false. The while loop terminates when it is equal 
to false, if it does indeed terminate. So we have two 
possible cases to prove this equality:


Case 1: The while loop never terminates.

We know that both pieces of code are syntactically identical
up until the end of the while loop. If the while loop does 
not terminate, it loops forever and continually calls C1, in 
both cases. If the while loop never terminates, nothing will 
be executed afterwards. So, both programs continually call C1 
forever in this case, making them semantically equivalent as 
well.

Case 2: The while loop terminates:
The while loop terminates if and only if E is false. So we 
assume that because both E’s and both C1’s in the left and 
right side of the equality are the same, that the while loop 
on both sides will execute C1 and terminate in the same amount 
of steps. So, again, semantically and syntactically speaking, 
the programs are identical up until the end of the while loop. 
The two programs differ only in the if then else statement 
following the while loop on the right side of the expression. 
Assuming the while loop terminates, it terminates when E is 
false, in both cases. So, the following if then else statement 
will only ever execute the else portion, because, when the while 
loop terminates, E will be false, so the "if" portion of the 
statement on the right side won’t execute. Because the "if" 
portion of the if then else statement will never execute, C2 
will always be called after the while loop in both the left and 
right expression. 
*)















(* James Caldwell                                                    
   University of Wyoming, Department of Computer Science, Laramie WY 
   COSC 4780 -- Principles of Programming Languages -- Spring 2016
*)

(* base code for hw6 *)

(* ====================================================================== *)
(* UTILITIES                                                              *)
(* ====================================================================== *)
 

(* compute the intersection of two lists *)

let list_intersection m n = 
  List.fold_right (fun x y -> if List.mem x n then x:: y else y) m [];;

(* ====================================================================== *)
(* ABSTRACT SYNTAX                                                        *)
(* ====================================================================== *)

(* ---------------------------------------------------------------------- *)
(* locations                                                              *)
(* ---------------------------------------------------------------------- *)

type loc =  Loc of int ;;

let string_of_loc l = 
  match l  with
   Loc i -> "Loc " ^ (string_of_int i)
;;

(* ---------------------------------------------------------------------- *)
(* expressions                                                            *)
(* ---------------------------------------------------------------------- *)

type expression =
    N of int
  | Deref of loc
  | Add of expression * expression
  | Neg of expression
  | Eq of expression * expression
;;

let rec string_of_expression e =
  match e with 
    N (i) -> "N " ^ (string_of_int i)
  | Deref (al) -> "Deref " ^ (string_of_loc al)
  | Add (e1,e2) -> "Add (" ^ (string_of_expression e1) ^ ", " ^ (string_of_expression e1) ^ ")"
  | Neg (e1) -> "Neg " ^ (string_of_expression e1)
  | Eq (e1,e2) -> "Eq  (" ^ (string_of_expression e1) ^ ", " ^ (string_of_expression e1) ^ ")"
;;


(* ---------------------------------------------------------------------- *)
(* commands                                                               *)
(* ---------------------------------------------------------------------- *)

type command = 
    Assign of loc * expression
  | Seq of command * command
  | If of expression * command * command
  | While of expression * command
  | Skip
;;

let rec string_of_command c = 
  match c with 
    Assign (al,e) -> "Assign (" ^ (string_of_loc al) ^ ", " ^ (string_of_expression e) ^ ")"
  | Seq (c1,c2) -> "Seq (" ^ (string_of_command c1) ^ ", " ^ (string_of_command c2) ^ ")"
  | If (e,c1,c2) -> "If (" ^ (string_of_expression e) ^ ", " ^ (string_of_command c1) ^ ", " ^ (string_of_command c2) ^ ")"
  | While (e,c1) -> "While (" ^ (string_of_expression e) ^ ", " ^ (string_of_command c1) ^ ")"
  | Skip  -> "Skip "
;;


(* ====================================================================== *)
(* ANALYSIS                                                               *)
(* ====================================================================== *)

(* ---------------------------------------------------------------------- *)
(* reports                                                                *)
(* ---------------------------------------------------------------------- *)

type report = Good | Possible of (command list);;

let union_reports r1 r2 = 
  match r1 with
      Good -> r2
    | Possible c1 -> (match r2 with
			  Good -> r1
			| Possible c2 -> Possible (c1 @ c2))
		      ;;

(* ---------------------------------------------------------------------- *)
(* the analysis                                                           *)
(* ---------------------------------------------------------------------- *)

(* implement the following functions *)
let rec exp_refs e = 
  match e with
    N (i) -> []
  | Deref (al) -> al::[]
  | Add (e1,e2) -> (exp_refs e1)@(exp_refs e2)
  | Neg (e1) -> exp_refs e1
  | Eq (e1,e2) -> (exp_refs e1)@(exp_refs e2)
;;

let rec active_locs c =  
  match c with
    Assign (al,e) -> al::[]
  | Seq (c1,c2) -> (active_locs c1)@(active_locs c2)
  | If (e,c1,c2) -> (active_locs c1)@(active_locs c2)
  | While (e,c1) -> (active_locs c1)
  | Skip  -> []
;;
        
let rec loops_analysis c =
  match c with
    Assign (al,e) -> Good
  | Seq (c1,c2) -> union_reports(loops_analysis c1) (loops_analysis c2)
  | If (e,c1,c2) ->  union_reports(loops_analysis c1) (loops_analysis c2)
  | While (e,c1) -> let intersection = 
                      list_intersection (active_locs c1) (exp_refs e) in
                    if intersection = [] then 
                      union_reports (Possible [c]) (loops_analysis c1)
                    else
                      union_reports (Good) (loops_analysis c1)  
                       
  | Skip  -> Good
;;  
