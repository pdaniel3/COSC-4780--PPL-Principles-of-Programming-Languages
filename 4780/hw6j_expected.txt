# #use "parser_base.ml";;
type 'a parser = P of (string -> ('a * string) list)
val parse : 'a parser -> string -> ('a * string) list = <fun>
val analyze : 'a parser -> string -> 'a = <fun>
val return : 'a -> 'a parser = <fun>
val bind : 'a parser -> ('a -> 'b parser) -> 'b parser = <fun>
val bind' : 'a parser -> 'b parser -> 'b parser = <fun>
val mzero : 'a parser = P <fun>
val mplus : 'a parser -> 'a parser -> 'a parser = <fun>
val failure : 'a parser = P <fun>
val char2str : char -> string = <fun>
val shead : string -> char = <fun>
val rest : string -> string = <fun>
val item : char parser = P <fun>
val choice : 'a parser -> 'a parser -> 'a parser = <fun>
val choices : 'a parser list -> 'a parser = <fun>
val chainl1 : 'a parser -> ('a -> 'a -> 'a) parser -> 'a parser = <fun>
val chainl : 'a parser -> ('a -> 'a -> 'a) parser -> 'a -> 'a parser = <fun>
val sat : (char -> bool) -> char parser = <fun>
val orP : ('a -> bool) -> ('a -> bool) -> 'a -> bool = <fun>
val isDigit : char -> bool = <fun>
val isLower : char -> bool = <fun>
val isUpper : char -> bool = <fun>
val isAlpha : char -> bool = <fun>
val isSpace : char -> bool = <fun>
val isWhiteSpace : char -> bool = <fun>
val digit : char parser = P <fun>
val lower : char parser = P <fun>
val upper : char parser = P <fun>
val letter : char parser = P <fun>
val char : char -> char parser = <fun>
val alphanum : char parser = P <fun>
val string : string -> string parser = <fun>
val many1 : char parser -> string parser = <fun>
val many : char parser -> string parser = <fun>
val list1P : 'a parser -> 'a list parser = <fun>
val listP : 'a parser -> 'a list parser = <fun>
val lowerIdent : string parser = P <fun>
val upperIdent : string parser = P <fun>
val ident : string parser = P <fun>
val nat : int parser = P <fun>
val int : int parser = P <fun>
val space : unit parser = P <fun>
val whiteSpace : unit parser = P <fun>
val token : 'a parser -> 'a parser = <fun>
val prefix : string -> unit parser = <fun>
val delimted_by : char -> string parser = <fun>
val tokenized_delimted_by : char -> string parser = <fun>
val identifier : string parser = P <fun>
val natural : int parser = P <fun>
val integer : int parser = P <fun>
val symbol : string -> string parser = <fun>
val wrapped : (unit -> 'a parser) -> 'a parser = <fun>
# #use "parser_ch1.ml";;
type 'a parser = P of (string -> ('a * string) list)
val parse : 'a parser -> string -> ('a * string) list = <fun>
val analyze : 'a parser -> string -> 'a = <fun>
val return : 'a -> 'a parser = <fun>
val bind : 'a parser -> ('a -> 'b parser) -> 'b parser = <fun>
val bind' : 'a parser -> 'b parser -> 'b parser = <fun>
val mzero : 'a parser = P <fun>
val mplus : 'a parser -> 'a parser -> 'a parser = <fun>
val failure : 'a parser = P <fun>
val char2str : char -> string = <fun>
val shead : string -> char = <fun>
val rest : string -> string = <fun>
val item : char parser = P <fun>
val choice : 'a parser -> 'a parser -> 'a parser = <fun>
val choices : 'a parser list -> 'a parser = <fun>
val chainl1 : 'a parser -> ('a -> 'a -> 'a) parser -> 'a parser = <fun>
val chainl : 'a parser -> ('a -> 'a -> 'a) parser -> 'a -> 'a parser = <fun>
val sat : (char -> bool) -> char parser = <fun>
val orP : ('a -> bool) -> ('a -> bool) -> 'a -> bool = <fun>
val isDigit : char -> bool = <fun>
val isLower : char -> bool = <fun>
val isUpper : char -> bool = <fun>
val isAlpha : char -> bool = <fun>
val isSpace : char -> bool = <fun>
val isWhiteSpace : char -> bool = <fun>
val digit : char parser = P <fun>
val lower : char parser = P <fun>
val upper : char parser = P <fun>
val letter : char parser = P <fun>
val char : char -> char parser = <fun>
val alphanum : char parser = P <fun>
val string : string -> string parser = <fun>
val many1 : char parser -> string parser = <fun>
val many : char parser -> string parser = <fun>
val list1P : 'a parser -> 'a list parser = <fun>
val listP : 'a parser -> 'a list parser = <fun>
val lowerIdent : string parser = P <fun>
val upperIdent : string parser = P <fun>
val ident : string parser = P <fun>
val nat : int parser = P <fun>
val int : int parser = P <fun>
val space : unit parser = P <fun>
val whiteSpace : unit parser = P <fun>
val token : 'a parser -> 'a parser = <fun>
val prefix : string -> unit parser = <fun>
val delimted_by : char -> string parser = <fun>
val tokenized_delimted_by : char -> string parser = <fun>
val identifier : string parser = P <fun>
val natural : int parser = P <fun>
val integer : int parser = P <fun>
val symbol : string -> string parser = <fun>
val wrapped : (unit -> 'a parser) -> 'a parser = <fun>
val locationP : loc parser = P <fun>
val symEq : string parser = P <fun>
val symPlus : string parser = P <fun>
val symAt : string parser = P <fun>
val symNot : string parser = P <fun>
val numP : expression parser = P <fun>
val derefP : expression parser = P <fun>
val notP : (unit -> expression parser) -> expression parser = <fun>
val eqP : (unit -> expression parser) -> expression -> expression parser =
  <fun>
val plusP : (unit -> expression parser) -> expression -> expression parser =
  <fun>
val expressionP : expression parser = P <fun>
val symSemi : string parser = P <fun>
val symAssign : string parser = P <fun>
val symIf : string parser = P <fun>
val symThen : string parser = P <fun>
val symElse : string parser = P <fun>
val symFi : string parser = P <fun>
val symWhile : string parser = P <fun>
val symDo : string parser = P <fun>
val symOd : string parser = P <fun>
val symCall : string parser = P <fun>
val assignP : command parser = P <fun>
val ifthenelseP : (unit -> command parser) -> command parser = <fun>
val whileP : (unit -> command parser) -> command parser = <fun>
val skipP : command parser = P <fun>
val seqP :
  (unit -> command parser) -> (unit -> command parser) -> command parser =
  <fun>
val commandP : command parser = P <fun>
# let test f  p = 
    f (analyze commandP p);;
  val test : (command -> 'a) -> string -> 'a = <fun>
# let test_exp_refs e = exp_refs (analyze expressionP e);;
val test_exp_refs : string -> loc list = <fun>
# test_exp_refs "@loc2 = 1";;
- : loc list = [Loc 2]
# test_exp_refs "(1 = 1) = ~(2 = 2)";;
- : loc list = []
# test_exp_refs "(1 = @loc1) = ~(2 = @loc2)";;
- : loc list = [Loc 1; Loc 2]
# test_exp_refs "@loc1 = @loc2";;
- : loc list = [Loc 1; Loc 2]
# test_exp_refs "@loc3 = @loc1";;
- : loc list = [Loc 3; Loc 1]
# test_exp_refs "@loc3 + @loc1";;
- : loc list = [Loc 3; Loc 1]
# test_exp_refs "@loc3 + 1";;
- : loc list = [Loc 3]
# test active_locs "loc1 := @loc1";;
- : loc list = [Loc 1]
# test active_locs "loc1 := @loc2";;
- : loc list = [Loc 1]
# test active_locs "loc1 := (-2); while ~(@loc1 = 1) do loc1:= @loc2 + 1 od";;
- : loc list = [Loc 1; Loc 1]
# test active_locs "loc1 := (-2); while ~(@loc1 = 1) do loc1:= @loc2 + 1; loc2 := @loc1 od";;
- : loc list = [Loc 1; Loc 1; Loc 2]
# test active_locs "if @loc1 = @loc2 then loc3 := @loc2 + 1 else loc4 := @loc 1 + 1 fi";;
- : loc list = [Loc 3; Loc 4]
# test loops_analysis "loc1 := @loc1";;
- : report = Good
# test loops_analysis "while (@loc1 = 0) do loc1 := @loc1 + 1 od";;
- : report = Good
# test loops_analysis "while (@loc2 = 0) do loc1 := @loc1 + 1 od";;
- : report =
Possible
 [While (Eq (Deref (Loc 2), N 0), Assign (Loc 1, Add (Deref (Loc 1), N 1)))]
# test loops_analysis "while (@loc2 = 0) do loc1 := @loc1 + 1 od; while (@loc2 = 1) do loc2 := 1 od";;
- : report =
Possible
 [While (Eq (Deref (Loc 2), N 0), Assign (Loc 1, Add (Deref (Loc 1), N 1)))]
# test loops_analysis "while (@loc2 = 0) do loc1 := @loc1 + 1 od; while (@loc2 = 1) do loc1 := 1 od";;
- : report =
Possible
 [While (Eq (Deref (Loc 2), N 0), Assign (Loc 1, Add (Deref (Loc 1), N 1)));
  While (Eq (Deref (Loc 2), N 1), Assign (Loc 1, N 1))]
# test loops_analysis "if (0=0) then while (@loc1 = 0) do loc1 := @loc1 + 1 od else skip fi";;
- : report = Good
# test loops_analysis "if (0=0) then skip else while (@loc1 = 0) do loc1 := @loc1 + 1 od  fi";;
- : report = Good
# test loops_analysis "if (0=0) then while (@loc2 = 0) do loc1 := @loc1 + 1 od else skip fi";;
- : report =
Possible
 [While (Eq (Deref (Loc 2), N 0), Assign (Loc 1, Add (Deref (Loc 1), N 1)))]
# test loops_analysis "if (0=0) then skip else while (@loc2 = 0) do loc1 := @loc1 + 1 od fi";;
- : report =
Possible
 [While (Eq (Deref (Loc 2), N 0), Assign (Loc 1, Add (Deref (Loc 1), N 1)))]
# 
